File: domain/src/main/java/com/beat/domain/member/domain/Member.java
package com.beat.domain.member.domain;

import java.time.LocalDateTime;

import com.beat.domain.BaseTimeEntity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Member extends BaseTimeEntity {

	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private Long id;

	@Column(nullable = false)
	private String nickname;

	@Column(nullable = true)
	private String email;

	@Column(nullable = true)
	private LocalDateTime deletedAt;

	@Column(name = "user_id", nullable = false)
	private Long userId;

	@Column(nullable = false)
	private Long socialId;  // 카카오 회원번호 저장

	@Enumerated(EnumType.STRING)
	private SocialType socialType;

	@Builder
	private Member(String nickname, String email, LocalDateTime deletedAt, Long userId, Long socialId,
		SocialType socialType) {
		this.nickname = nickname;
		this.email = email;
		this.deletedAt = deletedAt;
		this.userId = userId;
		this.socialId = socialId;
		this.socialType = socialType;
	}

	public static Member create(
		final String nickname,
		final String email,
		final Long userId,
		final Long socialId,
		final SocialType socialType
	) {
		return Member.builder()
			.nickname(nickname)
			.email(email)
			.userId(userId)
			.socialId(socialId)
			.socialType(socialType)
			.build();
	}
}

File: domain/src/main/kotlin/com/beat/domain/member/domain/Member.kt
package com.beat.domain.member.domain

import com.beat.domain.user.domain.Users
import java.time.LocalDateTime

@ConsistentCopyVisibility
data class Member private constructor(
    private val memberId: Id?,
    val nickname: String,
    val email: String?,
    val deletedAt: LocalDateTime?,
    private val linkedUserId: Users.Id,
    val socialId: Long,
    val socialType: SocialType,
) {
    fun getId(): Long? = memberId?.value

    fun getUserId(): Long = linkedUserId.value

    @JvmInline
    value class Id private constructor(val value: Long) {
        companion object {
            @JvmStatic
            fun from(value: Long): Id = Id(value)

            @JvmStatic
            fun fromNullable(value: Long?): Id? = value?.let(::from)
        }
    }

    companion object {
        @JvmStatic
        fun create(
            nickname: String,
            email: String?,
            userId: Long,
            socialId: Long,
            socialType: SocialType,
        ): Member = Member(
            memberId = null,
            nickname = nickname,
            email = email,
            deletedAt = null,
            linkedUserId = Users.Id.from(userId),
            socialId = socialId,
            socialType = socialType,
        )

        @JvmStatic
        fun rehydrate(
            id: Long?,
            nickname: String,
            email: String?,
            deletedAt: LocalDateTime?,
            userId: Long,
            socialId: Long,
            socialType: SocialType,
        ): Member = Member(
            memberId = Id.fromNullable(id),
            nickname = nickname,
            email = email,
            deletedAt = deletedAt,
            linkedUserId = Users.Id.from(userId),
            socialId = socialId,
            socialType = socialType,
        )
    }
}
